File: day12_part1_2022.java

import java.io.File;
import java.io.FileNotFoundException;
import java.util.PriorityQueue;
import java.util.Scanner;

public class Solution {

    public static void main(String[] args) {
        char[][] grid = new char[1000][1000];
        int[][] dists = new int[1000][1000];
        Point start = new Point(0, 0);
        Point end = new Point(0, 0);
        Point[] as = new Point[1000];
        int y = 0;
        try {
            Scanner scanner = new Scanner(new File("input.txt"));
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                for (int x = 0; x < line.length(); x++) {
                    char b = line.charAt(x);
                    Point p = new Point(x, y);
                    grid[p.x][p.y] = b;
                    if (b == 'S') {
                        start = p;
                    } else if (b == 'E') {
                        end = p;
                    } else if (b == 'a') {
                        as[y] = p;
                    }
                }
                y++;
            }
            grid[start.x][start.y] = 'a';
            grid[end.x][end.y] = 'z';

            dists = djikstra(grid, end);

            int l = dists[start.x][start.y];
            System.out.println(l);

        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    static class Point {
        int x;
        int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        Point add(Point other) {
            return new Point(x + other.x, y + other.y);
        }
    }

    static class Item {
        Point obj;
        int priority;

        Item(Point obj, int priority) {
            this.obj = obj;
            this.priority = priority;
        }
    }

    static Point[] Neighbors4 = {new Point(0, 1), new Point(0, -1), new Point(1, 0), new Point(-1, 0)};

    static int[][] djikstra(char[][] grid, Point end) {
        PriorityQueue<Item> pq = new PriorityQueue<>((a, b) -> b.priority - a.priority);
        int[][] dist = new int[1000][1000];
        pq.add(new Item(end, 0));
        dist[end.x][end.y] = 0;
        while (!pq.isEmpty()) {
            Point curr = pq.poll().obj;
            for (Point n : Neighbors4) {
                Point next = curr.add(n);
                if (next.x < 0 || next.x >= 1000 || next.y < 0 || next.y >= 1000 || grid[next.x][next.y] == '\u0000') {
                    continue;
                }
                if (grid[curr.x][curr.y] - grid[next.x][next.y] > 1) {
                    continue;
                }
                int nextdist = dist[curr.x][curr.y] + 1;
                if (dist[next.x][next.y] == 0 || nextdist < dist[next.x][next.y]) {
                    dist[next.x][next.y] = nextdist;
                    pq.add(new Item(next, nextdist));
                }
            }
        }
        return dist;
    }
}

File: c/day12_part1_2022.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#define MAX_GRID_SIZE 200
#define MAX_NODES MAX_GRID_SIZE * MAX_GRID_SIZE

typedef struct {
    int x, y;
} Point;

typedef struct {
    Point p;
    int priority;
} Item;

typedef struct {
    Item* data;
    int size;
    int capacity;
} PriorityQueue;

int grid[MAX_GRID_SIZE][MAX_GRID_SIZE];
int dist[MAX_GRID_SIZE][MAX_GRID_SIZE];
int rows, cols;

PriorityQueue* createPriorityQueue(int capacity) {
    PriorityQueue* pq = (PriorityQueue*)malloc(sizeof(PriorityQueue));
    pq->data = (Item*)malloc(sizeof(Item) * capacity);
    pq->size = 0;
    pq->capacity = capacity;
    return pq;
}

void freePriorityQueue(PriorityQueue* pq) {
    free(pq->data);
    free(pq);
}


void swap(Item* a, Item* b) {
    Item temp = *a;
    *a = *b;
    *b = temp;
}

void heapifyDown(PriorityQueue* pq, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < pq->size && pq->data[left].priority > pq->data[largest].priority) {
        largest = left;
    }

    if (right < pq->size && pq->data[right].priority > pq->data[largest].priority) {
        largest = right;
    }

    if (largest != i) {
        swap(&pq->data[i], &pq->data[largest]);
        heapifyDown(pq, largest);
    }
}

void heapifyUp(PriorityQueue* pq, int i) {
    while(i > 0 && pq->data[(i - 1) / 2].priority < pq->data[i].priority) {
      swap(&pq->data[i], &pq->data[(i - 1) / 2]);
      i = (i - 1) / 2;
    }
}

void push(PriorityQueue* pq, Point p, int priority) {
    if (pq->size == pq->capacity) {
        pq->capacity *= 2;
        pq->data = (Item*)realloc(pq->data, sizeof(Item) * pq->capacity);
    }
    pq->data[pq->size].p = p;
    pq->data[pq->size].priority = priority;
    heapifyUp(pq, pq->size);
    pq->size++;
}

Item pop(PriorityQueue* pq) {
    Item item = pq->data[0];
    pq->data[0] = pq->data[pq->size - 1];
    pq->size--;
    heapifyDown(pq, 0);
    return item;
}


int isValid(int x, int y) {
    return x >= 0 && x < cols && y >= 0 && y < rows;
}

int dx[] = {0, 0, 1, -1};
int dy[] = {1, -1, 0, 0};

int dijkstra(Point end, int start_x, int start_y) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
           dist[i][j] = INT_MAX;
        }
    }
    PriorityQueue* pq = createPriorityQueue(MAX_NODES);
    dist[end.y][end.x] = 0;
    push(pq, end, 0);

    while (pq->size > 0) {
        Item current_item = pop(pq);
        Point curr = current_item.p;
        int curr_dist = current_item.priority;

        if (curr_dist > dist[curr.y][curr.x]) {
          continue;
        }

        for (int i = 0; i < 4; i++) {
            int next_x = curr.x + dx[i];
            int next_y = curr.y + dy[i];

            if (isValid(next_x, next_y) &&
               grid[curr.y][curr.x] - grid[next_y][next_x] <= 1)
            {
              int next_dist = curr_dist + 1;
               if (next_dist < dist[next_y][next_x]) {
                  dist[next_y][next_x] = next_dist;
                  push(pq, (Point){next_x, next_y}, next_dist);
              }
            }
        }
    }
    freePriorityQueue(pq);
    return dist[start_y][start_x];
}

int main() {
    FILE* file = fopen("input.txt", "r");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    char line[MAX_GRID_SIZE];
    rows = 0;
    Point start, end;
    int start_x, start_y;
    while (fgets(line, sizeof(line), file)) {
        cols = strlen(line) - 1;
        for (int x = 0; x < cols; x++) {
            grid[rows][x] = line[x];
            if (line[x] == 'S') {
                start_x = x;
                start_y = rows;
            } else if (line[x] == 'E') {
                end.x = x;
                end.y = rows;
            }
        }
         rows++;
    }
    fclose(file);
    grid[start_y][start_x] = 'a';
    grid[end.y][end.x] = 'z';


    int shortest_path = dijkstra(end,start_x, start_y);

    printf("%d\n", shortest_path);
    return 0;
}
